### PR TITLE
add a 'status' command which prints the lag (& in JSON)

### PR DESCRIPTION
**Pull request by amandasaurus**
This patch adds an `info` command which prints readable summary of the state of replication:

```
$ osm2pgsql-replication info -U osm -d gis
2021-08-04 14:12:28 [INFO]: Using replication service 'https://download.geofabrik.de/europe-updates/'. Current sequence 3052 (2021-08-03T20:21:39Z).
2021-08-04 14:12:28 [INFO]: Server is at sequence 3052 (2021-08-03T20:21:39Z)
2021-08-04 14:12:28 [INFO]: Server's most recent data is 17 hour(s) 50 minutes(s) old
2021-08-04 14:12:28 [INFO]: Database is up to date with server
2021-08-04 14:12:28 [INFO]: Database's most recent data is 17 hour(s) 50 minutes(s) old
```

with the `--json` option, the output is in JSON format, which is useful for writing scripts to detect if the replication has stopped, or to graph how far behind the server is:

`$ osm2pgsql-replication info -U osm -d gis --json`:
```json
{
  "server": {
    "base_url": "https://download.geofabrik.de/europe-updates/",
    "sequence": 3052,
    "timestamp": "2021-08-03T20:21:39Z",
    "age": {
      "seconds": 64920,
      "text": {
        "en": "18 hour(s) 2 minutes(s)"
      }
    }
  },
  "local": {
    "sequence": 3052,
    "timestamp": "2021-08-03T20:21:39Z",
    "age": {
      "seconds": 64920,
      "text": {
        "en": "18 hour(s) 2 minutes(s)"
      }
    }
  },
  "diff": {
    "sequence": 0,
    "timestamp": {
      "seconds": 0,
      "text": {
        "en": "0 minutes(s)"
      }
    }
  }
}
```

### CODE REVIEW

**Comment by amandasaurus** [comment]
Please don't merge this now. I have unintentionally added something to the man page for `--post-processing`.

**Comment by lonvia** [comment]
> Please don't merge this now. I have unintentionally added something to the man page for --post-processing.

That part is okay. The man page wasn't regenerated last time.

**Comment by joto** [comment]
How about naming this `status` instead of `info`? `info` is very close to `init`, especially when some kind of command line completion is involved. Also `status` is more in line with what, for instance, `systemctl` or `git` use.

**Comment by amandasaurus** [comment]
I have renamed it to `status` (& force pushed the branch to update this PR)

File: scripts/osm2pgsql-replication
Comment (lonvia) [inline scripts/osm2pgsql-replication:140]: I could swear the initial version had more documentation in the man page. Extended documentation (including a description of the json output should be added here and will then automatically appear in the man page.

Also: please finish the sentence with a full stop.

File: scripts/osm2pgsql-replication
Comment (lonvia) [inline scripts/osm2pgsql-replication:446]: Full stop missing in help text.

File: scripts/osm2pgsql-replication
Comment (lonvia) [inline scripts/osm2pgsql-replication:57]: minutes -> minute

File: scripts/osm2pgsql-replication
Comment (lonvia) [inline scripts/osm2pgsql-replication:180]: `now()` should be called only once or you might end up with odd discrepancies for the age.

**Review by lonvia — CHANGES_REQUESTED** [review]
This is okay in principle but I would like to discuss the JSON output. I find the duplicated information in the age fields rather strange. If you have a computer-readable format, then providing the computer-readable output should be enough. Pretty-printing should be done on the client side. So I would just do `age: <age in seconds>`. Same goes for the `diff` section. It's trivial for a client to subtract sequence numbers or age seconds. No need to duplicate that on the server side.

File: scripts/osm2pgsql-replication
Comment (amandasaurus) [reply scripts/osm2pgsql-replication:140]: You can see the original version here. It had the same (low level of) detail https://github.com/openstreetmap/osm2pgsql/commit/eb4c79c4625c9fc384a3e1701c9f3a4fd07194f6

**Comment by amandasaurus** [comment]
I am working on your feedback. In order to track things, I will not rebase now. (Maybe we can rebase at end)

**Comment by amandasaurus** [comment]
> I would like to discuss the JSON output

How about this for an output format?

```{
  "server": {
    "base_url": "https://download.geofabrik.de/europe-updates/",
    "sequence": 3052,
    "timestamp": "2021-08-03T20:21:39Z",
    "age_sec": 64920
  },
  "local": {
    "sequence": 3052,
    "timestamp": "2021-08-03T20:21:39Z",
    "age_sec":  64920
  }
}
```

I prefer `age_secs` rather than just `age`, so that people will know the units (seconds) without having to look it up. Someone might think the `age` is the age in sequence numbers (for example).

`timestamp` is the timestamp of the server/local db, and `age_sec` is the difference between that timestamp, and now. This does mean `age_sec` is duplicating information that could be derived from the `timestamp` field. However given that parsing ISO timestamps & doing date diffs takes a little more than zero work, I think it is beneficial for the user to do the work in the `status` command.

If that output format is OK, I can implement it.

File: scripts/osm2pgsql-replication
Comment (amandasaurus) [reply scripts/osm2pgsql-replication:169]: Done. I had presumed that pyosmium would raise an exception. 😉

File: scripts/osm2pgsql-replication
Comment (amandasaurus) [reply scripts/osm2pgsql-replication:212]: For the JSON output, it does use `print(…)`. For the default "human readable format", it uses `LOG.info/fatal`

**Comment by lonvia** [comment]
That json format looks good to me. One more suggestion: how about adding a status field that is also returned in the case of an error? It would make it easier to check for success. Alternatively, you could always return the error code and make it 0 on success.

**Comment by amandasaurus** [comment]
I've changed the json output as requested.

I've added a `status` key to the JSON output, which is always the same as the exit code. so `0` is "OK", and then errors are `1`, `2` or `3`.

I've written help docs & updated man page

If you don't want any more changes, should I rebase again master? or squash? or just leave as it (i.e. you can merge it :) )

File: scripts/osm2pgsql-replication
Comment (lonvia) [reply scripts/osm2pgsql-replication]: This is the last bit where we disagree. Given that the output is the direct product of the command, it should go directly to stdout and not through the logging subsystem. So my preferred handling would be to leave `LOG.fatal()` in place but replace the `LOG.info()` with `print()`.

**Comment by lonvia** [comment]
Merging the other PR now causes conflicts for the man page. Can you please rebase? Stashing would not be a bad idea either but I leave that for your personal preferences.

**Comment by amandasaurus** [comment]
Squashed it all into one commit off current `master`.

File: scripts/osm2pgsql-replication
Comment (amandasaurus) [reply scripts/osm2pgsql-replication]: My apologies, I misunderstood you. I thought you were talking about the JSON. I have made the change you have requested.

**Review by lonvia — APPROVED** [review]

**Comment by joto** [comment]
I wonder if there is value in having this command optionally not contacting the remote server. If you only want to check what the status of the local database is, not contacting the server would eliminate a potential source of problems. Might be useful when this is run in some kind of monitoring environment. Then again, maybe not. Could be an option `--local-only`, or the other way around `--remote` or something like it?

**Comment by amandasaurus** [comment]
> I wonder if there is value in having this command optionally not contacting the remote server. 

That functionality is not needed for my use case, so I have not added it. It might be useful for someone. Will yous only merge this PR if I add that functionality? 🙂

**Comment by joto** [comment]
Fair enough. We'll leave this as "future work" then if and when somebody needs it.

**Pull request merged**
